# Problem Set 3

Related lectures: [Collections](https://datascience.quantecon.org/../python_fundamentals/collections.html) and [Control Flow](https://datascience.quantecon.org/../python_fundamentals/control_flow.html). Feel free to apply the linear algebra in Numpy arrays if you find it helpful.


* Group Members
    * Cathy Liu (94353109)
    * Jossie Jiang (48372874)

* Contribution Clarification
    * Cathy Liu and Jossie Jiang each did their own problem sets and then compared answers. Since the answers mostly are the same, we just kept the way we each derived our answers. Except the qeustion 3, Cathy and Jossie worked together and finnally agreed with each other.

## Question 1-3

Consider a bond that pays a \$500 coupon once every quarter.

It pays in March, June, September, and December.

It promises to do so for 10 years after you purchase it, from January 2022 to December 2031.

You discount the future at the rate of $r = 0.005$ **per month**.

### Question 1

How much do you value the asset in January 2022?

In [1]:
discount_factor = 0.005*3
payoff = 500
pv = 0
for i in range(0, 40):
    pv = (1/(1+discount_factor)**(i+1))*payoff + pv

The present value of the asset in January 2022 is 14957.92

### Question 2

Consider a different asset that pays a lump sum at its expiration date rather than a quarterly coupon of \$500 dollars.

How much would this asset need to pay on December 2031 for the two assets to be equally valued?

In [2]:
lump_sum = round(pv*(discount_factor+1)**40,2)
print(lump_sum)

27133.95


There will be lump sum of 27133.95 being needed to pay on Dec. 2031

### Question 3

If you already knew that your discount rate would change annually according to the
table below, at what price would you value the quarterly asset (from the main text) in January 2022?

*Hint*: There are various ways to do this… One way might include a zipped loop for years and a
second loop for months.

*More Challenging*: Can you create the list of interest rates without calculating each year individually?

|Year|Discount Rate|
|:------------------------------------------------:|:------------------------------------------------:|
|2022|0.005|
|2023|0.0047|
|2024|0.0044|
|2025|0.0041|
|2026|0.0038|
|2027|0.0035|
|2028|0.0032|
|2029|0.0029|
|2030|0.0026|
|2031|0.0023|

Hint: create appropriate collections typing from the data directly in the code.  You cannot parse the
text table directly.

In [3]:
rates = [0.005, 0.0047, 0.0044, 0.0041, 0.0038, 0.0035, 0.0032, 0.0029, 0.0026, 0.0023]

rates = [i * 3 for i in rates]
rates = [j for j in rates for _ in range(0,4)]

npv = 0
payoff = 500

for i, r in enumerate(rates):
    nper = i+1
    npv += payoff/((1+r)**nper)

print(round(npv,2))

16527.71


the quarterly asset is valued at 16527.71 in Jan. 2022

## Questions 4-5

Companies often invest in training their employees to raise their
productivity. Economists sometimes wonder why companies
spend money on training employees when this incentivizes other companies to poach
their employees with higher salaries since the employees gain human capital from training.

Imagine it costs a company 30,000 dollars to teach their employees Python in January, but it also raises their output by 2,500 dollars per month from February on. The company discounts the future at rate of $ r = 0.01 $ per month.

### Question 4

For how many full months does an employee need to stay at a company for that company to make a profit for paying for their employees’ Python training?

In [4]:
cost = 30000
output_inc = 2500
r = 0.01
profit = 0
i = 1
while(cost > profit):
    profit = profit + (1/(1+r)**i)*output_inc
    i = i+1
print(i)

14


At least 14 full months an employee needs to stay at a company for that company to make a profit for paying for the Python training.

### Question 5

Imagine that 80% of the employees stay for 9 months and 20% of the employees stay
for 15 months. Is it worth it for the company to invest in employee Python training?

In [5]:
profit = 0
for i in range(0, 9):
    profit = profit + (1/(1+r)**(i+1))*output_inc*0.8
for i in range(0, 15):
    profit = profit + (1/(1+r)**(i+1))*output_inc*0.2
print(round(profit,2))

24064.56


Because under this circumstance, the profit generated will be 24064.56 which is smaller than 30000. So it's not worthy for the company to invest in employee Python training.